velox/dwio/common/ScanSpec.h

```diff
@@ -198,13 +212,15 @@ class ScanSpec {
 
   /// Returns the ScanSpec corresponding to 'name'. Creates it if needed without
   /// any intermediate level.
-  ScanSpec* getOrCreateChild(const std::string& name);
+  ScanSpec* getOrCreateChild(const std::string& name, bool isTempNode = false);
 
   // Returns the ScanSpec corresponding to 'subfield'. Creates it if
   // needed, including any intermediate levels. This is used at
   // TableScan initialization to create the ScanSpec tree that
   // corresponds to the ColumnReader tree.
-  ScanSpec* getOrCreateChild(const Subfield& subfield);
+  ScanSpec* getOrCreateChild(const Subfield& subfield, bool isTempNode = false);
+
+  void deleteTempNodes();
```

Comment:
Since all the temp nodes are top level, can we do it less intrusively by keeping a list of temp node/filter names inside `IcebergSplitReader` and remove them after we've done with it?  Then we don't need the extra states of `isTempNode_` and `hasTempFilter_` here.

Reply:
@Yuhta Thanks for the suggestion. Yes in our last discussion we thought we could just add/remove top level columns, but when I looked again I found it might be necessary to still mark the nodes with these temp tags, because the Iceberg delete file may contain subfields. For example, this query
```
-- create table t1 (c_row(c1 integer, c2 integer, c3 integer), c_char char);
-- insert some rows

select c_char
from t1 
where c_row.c2 > 2;
```
The base ScanSpec would contain the c_row child, which in turn has 3 children for c1, c2 and c3. In this case, all the three subfields would be null constants, but only c2 node would have filter `c_row.c2 > 2,` while c1 and c3 node don't have any filters. Now, if the Iceberg equality delete file contains the predicate `c_row.c1 = 1 && c_row.c2 IN {2,3}`, then we need to remove these values by adding a filter `c_row.c1<>1` for the c_row.c1 node, and merge the filter `c_row.c2<>2 && c_row.c2<>3` with existing filter `c_row.c2 > 2` for the c_row.c2 node, so it becomes 'c_row.c2 > 3`.  WHen this split finishes, we need to remove the filter `c_row.c1<>1`  from c_row.c1, and restore the filter for c_row.c2 to its original state, but not delete the whole c_row column from the ScanSpec. Therefore we need to tag the nodes in the ScanSpec to check if it's temp node and has temp filters. Do you think this makes sense?